ktor-network/common/src/io/ktor/network/sockets/SocketBase.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2014-2025 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.network.sockets
+
+import io.ktor.network.selector.*
+import io.ktor.utils.io.*
+import kotlinx.atomicfu.*
+import kotlinx.coroutines.*
+import kotlinx.io.*
+import kotlin.coroutines.*
+
+internal abstract class SocketBase(
+    parent: CoroutineContext
+) : ReadWriteSocket, SelectableBase(), CoroutineScope {
+
+    private val closeFlag = atomic(false)
+
+    private val readerJob = atomic<ReaderJob?>(null)
+
+    private val writerJob = atomic<WriterJob?>(null)
+
+    override val socketContext: CompletableJob = Job(parent[Job])
+
+    override val coroutineContext: CoroutineContext
+        get() = socketContext
+
+    override fun dispose() {
+        close()
+    }
+
+    override fun close() {
+        if (!closeFlag.compareAndSet(false, true)) return
+
+        readerJob.value?.channel?.close()
+        writerJob.value?.cancel()
+        checkChannels()
+    }
+
+    final override fun attachForReading(channel: ByteChannel): WriterJob {
+        return attachFor("reading", channel, writerJob) {
+            attachForReadingImpl(channel)
+        }
+    }
+
+    final override fun attachForWriting(channel: ByteChannel): ReaderJob {
+        return attachFor("writing", channel, readerJob) {
+            attachForWritingImpl(channel)
+        }
+    }
+
+    abstract fun attachForReadingImpl(channel: ByteChannel): WriterJob
+    abstract fun attachForWritingImpl(channel: ByteChannel): ReaderJob
+
+    private inline fun <J : ChannelJob> attachFor(
+        name: String,
+        channel: ByteChannel,
+        ref: AtomicRef<J?>,
+        producer: () -> J
+    ): J {
+        if (closeFlag.value) {
+            val e = IOException("Socket closed")
```

Comment:
cause

Reply:
should we use cancellation exception as a cause?

Reply:
@e5l I moved this from JVM code to common code and changed it from `ClosedChannelException` to `IOException` so it works in common. I think `IOException` is better that `CancellationException` as reading/writing on a closed socket seems like an error. A `CancellationException` would just cause it to be ignored. What do you think is best?